WordPress/wporg-mu-plugins
# Add wporg/time block

Closes #413 

This PR reimplements the existing [[time] shortcode](https://meta.trac.wordpress.org/browser/sites/trunk/wordpress.org/public_html/wp-content/mu-plugins/pub/wporg-time-shortcode.php) as a Gutenberg format.

The block uses the Format API so that the time can be applied to a string within a parent block, eg. a paragraph.

#### Notable changes from the existing shortcode implementation:

1. The reformatting is achieved by filtering and modifying `the_content` rather than the shortcode content
2. The time parsing (`parse_time`) is separated from the dom manipulation (`transform_time_blocks`)
3. The time parsing is always relative to the post date, there is no [attribute support](https://meta.trac.wordpress.org/browser/sites/trunk/wordpress.org/public_html/wp-content/mu-plugins/pub/wporg-time-shortcode.php#L53)
4. The frontend `time_converter_script` no longer has a jQuery dependency

### Screenshots

| Editor | Frontend |
|-|-|
| ![Screenshot 2023-07-11 at 1 22 27 PM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/d359f2a6-0c1b-4741-b36d-53b679d35b8e) | ![Screenshot 2023-07-11 at 5 44 40 PM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/b977ed6b-5993-464a-b1bc-3cd8a92f4eb4) | 

### Testing

1. Create a post, add various valid (eg. 'Tuesday, April 5th, at 15:00 UTC') and invalid datetime descriptions, and format as a `Time` using the option in the rich text toolbar 
2. Save and view on the frontend
3. The date displayed should be reformatted to your timezone
4. Test various styles
5. Change the post date and check that the displayed time changes too
6. Add multiple Times and check that the JS is only added to the page once
7. Test cross browser and [include Safari](https://github.com/WordPress/wporg-mu-plugins/pull/414/files#diff-56e893c6923a0a2c2b2277cfea55838a748dc980d25fb141597e82c8b4858b51R137)
